### ISSUE TITLE
ML - 성규

### ISSUE THREAD

**Issue by Seonggyu-Han** [skn-ai14-250409/SKN14-Final-2Team#41]
## ✔️ 상세 내용

> 구현 범위, 작업 방법, 조건, 진행 일자 등을 상황에 맞게 구체적으로 작성

## 🎯 이유/목적/배경

> 이 작업이 필요한 배경, 비즈니스/기술적 이유

## 📆 할 일

- [ ] (본인만의 체크리스트)
- [ ] (하나라도 필수 작성)

## 🎁 작업 결과

> 구현 완료된 결과(글/사진/그림 등), 장애사항도 있었다면 같이 작성

## 🔗 추가 정보/참고 자료

> 관련 문서, 스크린샷, 링크 등 첨부

**Comment by Seonggyu-Han** [comment]
<img width="688" height="270" alt="Image" src="https://github.com/user-attachments/assets/ddf03c71-c073-4d63-8874-870eba400e8f" />

목표 f1 micro score인 0.6에 미치지 못함.

<성능 향상 위한 수정 사항>


[토큰 길이·스케줄 튜닝]
1) max_len 변경
```
def tokenize(batch):
    return tok(
        batch[USE_TEXT_COL],
        truncation=True,
        padding="max_length",
        max_length=256 -> 512         #### description 끊기지 않도록 증가
    )
```

2) TrainingArguments 블록 수정
```
from transformers import TrainingArguments

args = TrainingArguments(
    OUTPUT_DIR,
    eval_strategy="epoch",
    save_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=16,
    gradient_accumulation_steps=2,            #### 유효 배치 16으로 증가
    num_train_epochs=2 -> 5 -> 6,              #### epoch 횟수 증가
    weight_decay=0.01,
    logging_steps=50,
    load_best_model_at_end=True,
    metric_for_best_model="micro_f1",
    greater_is_better=True,
    warmup_ratio=0.06,                              #### 워밍업 추가
    lr_scheduler_type="linear",                   #### 추가
    bf16=True if torch.cuda.is_available() else False,                          ####fp -> bf
    gradient_checkpointing=True
)

# gradient_checkpointing 사용 시 권장 (경고 방지)
model.config.use_cache = False                           ####

```

[불균형 보정: pos_weight 적용]

1) 불균형 보정: pos_weight 적용

라벨 희소/불균형 때문에 Macro-F1이 낮게 나온 상태야. BCEWithLogitsLoss에 pos_weight=(N−P)/P를 주면 희귀 라벨 가중이 들어가서 Micro/Macro 둘 다 개선.
```
# Y_tr는 MultiLabelBinarizer로 만든 (n_samples, n_labels)
pos = Y_tr.sum(axis=0)                     # 각 라벨의 양성 개수
neg = len(Y_tr) - pos
pos_weight = (neg / np.clip(pos, 1, None)).astype("float32")

from transformers import Trainer
import torch

class WeightedBCETrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False):
        labels = inputs["labels"]
        out = model(input_ids=inputs["input_ids"],
                    attention_mask=inputs["attention_mask"])
        loss = torch.nn.BCEWithLogitsLoss(
            pos_weight=torch.tensor(pos_weight, device=out.logits.device)
        )(out.logits, labels)
        return (loss, out) if return_outputs else loss
```

2) trainer = Trainer(...) 를 아래로 교체

```
trainer = WeightedBCETrainer(
    model=model,
    args=args,
    train_dataset=train_ds,
    eval_dataset=valid_ds,
    tokenizer=tok,
    compute_metrics=compute_metrics
)
```

**Comment by Seonggyu-Han** [comment]
코드 상 어떤 부분이 치명적인 성능 저하를 일으키고 있는지, 어떤 부분을 수정해야 효과적으로 성능이 상승할지 서칭 한 결과.

1) (치명적) 멀티라벨 분할이 사실상 “비-계층” 분할임 → Iterative Stratification으로 교체

문제)
현재 분할은 stratify=(Y.sum(axis=1) > 0)라서 “라벨이 1개 이상이냐”만 맞춥니다. 라벨별/라벨동시출현 분포가 검증셋에 유지되지 않아 평가 편차·학습 불안정이 큼.(희소 라벨 특히 치명적) .

개선)
멀티라벨에서는 iterative stratification(Sechidis et al.)을 써야 각 라벨 및 공발생 분포가 train/valid에 비슷하게 유지.

패치 코드 (xlm_learn.py의 train/valid split 대체)

```
# pip install iterative-stratification
from iterstrat.ml_stratifiers import MultilabelStratifiedShuffleSplit

msss = MultilabelStratifiedShuffleSplit(n_splits=1, test_size=0.1, random_state=42)
idx_tr, idx_va = next(msss.split(np.arange(len(df_trainable)), Y))

```

<img width="673" height="474" alt="Image" src="https://github.com/user-attachments/assets/6d731d1f-34c7-4ebb-9553-31464e79c290" />

== threshold 0.5 ==
              precision    recall  f1-score   support

          그린       0.23      0.39      0.29       152
         라벤더       0.00      0.00      0.00         0
          로즈       0.30      0.62      0.40        78
         머스크       0.20      0.39      0.26        46
          모스       0.12      0.22      0.16       125
        바이올렛       0.30      1.00      0.46         3
         스위트       0.00      0.00      0.00         0
        스파이시       0.07      0.12      0.09        25
        시트러스       0.33      0.53      0.41       498
        아로마틱       0.30      0.59      0.40       217
        아이리스       0.28      0.60      0.38        15
          앰버       0.45      0.78      0.57       895
          우디       0.56      0.72      0.63       933
        튜베로즈       0.32      0.55      0.40        11
         프레쉬       0.00      0.00      0.00        11
         프루티       0.33      0.60      0.43       458
         플로랄       0.67      0.76      0.71      1053

   micro avg       0.44      0.67      0.53      4520
   macro avg       0.26      0.46      0.33      4520
weighted avg       0.47      0.67      0.54      4520
 samples avg       0.47      0.69      0.53      4520

== per-label thresholds ==
              precision    recall  f1-score   support

          그린       0.26      0.38      0.31       152
         라벤더       0.00      0.00      0.00         0
          로즈       0.36      0.60      0.45        78
         머스크       0.24      0.35      0.28        46
          모스       0.17      0.17      0.17       125
        바이올렛       0.43      1.00      0.60         3
         스위트       0.00      0.00      0.00         0
        스파이시       0.10      0.12      0.11        25
        시트러스       0.37      0.47      0.41       498
        아로마틱       0.35      0.51      0.42       217
        아이리스       0.36      0.60      0.45        15
          앰버       0.44      0.83      0.57       895
          우디       0.54      0.77      0.63       933
        튜베로즈       0.40      0.55      0.46        11
         프레쉬       0.02      0.09      0.04        11
         프루티       0.38      0.49      0.43       458
         플로랄       0.62      0.86      0.72      1053

   micro avg       0.46      0.68      0.55      4520
   macro avg       0.30      0.46      0.36      4520
weighted avg       0.46      0.68      0.55      4520
 samples avg       0.49      0.71      0.55      4520


유의미한 차이 X

**Comment by Seonggyu-Han** [comment]
description -> main accord 결론 내리는 학습
: description에서 정규화를 통해 광고성/메타/브랜드 정보 등 불필요한 요소들을 제거

**Comment by Seonggyu-Han** [comment]
stopwords를 제거하지 않고, 브랜드와 향수명, 광고/메타 정보만 삭제, VotingClassifier 분류기 적용.



stopwords를 제거하지 않고, 브랜드와 향수명, 광고/메타 정보만 삭제, VotingClassifier 분류기 적용.
